prow/test/integration/test/sinker_test.go

```diff
@@ -0,0 +1,178 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	prowjobv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestDeletePod(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name       string
+		hasCR      bool
+		wantDelete bool
+	}{
+		{
+			"running-pod",
+			true,
+			false,
+		},
+		{
+			"orphaned-pod",
+			false,
+			true,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			clusterContext := getClusterContext()
+			t.Logf("Creating client for cluster: %s", clusterContext)
+
+			kubeClient, err := NewClients("", clusterContext)
+			if err != nil {
+				t.Fatalf("Failed creating clients for cluster %q: %v", clusterContext, err)
+			}
+
+			ctx := context.Background()
+
+			prowjob := prowjobv1.ProwJob{
+				ObjectMeta: v1.ObjectMeta{
+					Annotations: map[string]string{
+						"prow.k8s.io/job": tt.name,
+					},
+					Labels: map[string]string{
+						"created-by-prow":  "true",
+						"prow.k8s.io/type": "periodic",
+						"name":             tt.name,
+					},
+					Name:      tt.name,
+					Namespace: defaultNamespace,
+				},
+				Spec: prowjobv1.ProwJobSpec{
+					Type:      prowjobv1.PeriodicJob,
+					Namespace: testpodNamespace,
+					Job:       tt.name,
+				},
+				Status: prowjobv1.ProwJobStatus{
+					State: prowjobv1.TriggeredState,
+				},
+			}
+			pod := corev1.Pod{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      tt.name,
+					Namespace: testpodNamespace,
+					Labels: map[string]string{
+						"name":            tt.name,
+						"created-by-prow": "true",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "busybox",
+							Image: "localhost:5000/busybox",
+							Args: []string{
+								"sleep",
+								"1000000",
+							},
+						},
+					},
+				},
+			}
+
+			t.Cleanup(func() {
+				kubeClient.Delete(ctx, &prowjob)
+				kubeClient.Delete(ctx, &pod)
+			})
+
+			if tt.hasCR {
+				t.Logf("Creating CRD: %s", tt.name)
+
+				if err := kubeClient.Create(ctx, &prowjob); err != nil {
+					t.Fatalf("Failed creating prowjob: %v", err)
+				}
+				t.Logf("Finished creating CRD: %s", tt.name)
+			}
+			// Create pod
+			t.Logf("Creating pod: %s", tt.name)
+			if err := kubeClient.Create(ctx, &pod); err != nil {
+				t.Fatalf("Failed creating pod: %v", err)
+			}
+			t.Logf("Finished creating pod: %s", tt.name)
+
+			// Make sure pod is running
```

Comment:
This races with sinker deleting the pod. To safely test the orphaned pod case, I'd expect a PJ to be created before the pod is created, wait for the pod to start, then delete the PJ to orphan the pod. That should prevent sinker from seeing an orphaned pod until after we've confirmed the pod was successfully created.

Reply:
ack, done